src/Plugin/Action/EntityFetchByField.php

```diff
@@ -0,0 +1,128 @@
+<?php
+
+/**
+ * @file
+ * Contains \Drupal\rules\Plugin\Action\EntityFetchByField.
+ */
+
+namespace Drupal\rules\Plugin\Action;
+
+use Drupal\Core\Entity\EntityManagerInterface;
+use Drupal\Core\Plugin\ContainerFactoryPluginInterface;
+use Drupal\rules\Core\RulesActionBase;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+
+/**
+ * Provides a 'Fetch entities by field' action.
+ *
+ * @Action(
+ *   id = "rules_entity_fetch_by_field",
+ *   label = @Translation("Fetch entities by property"),
+ *   category = @Translation("Entity"),
+ *   context = {
+ *     "type" = @ContextDefinition("string",
+ *       label = @Translation("Entity type"),
+ *       description = @Translation("Specifies the type of the entity that should be fetched."),
+ *     ),
+ *     "property" = @ContextDefinition("string",
+ *       label = @Translation("Property"),
+ *       description = @Translation("The property by which the entity is to be selected.."),
+ *     ),
+ *     "value" = @ContextDefinition("any",
+ *       label = @Translation("Value"),
+ *       description = @Translation("The property value of the entity to be fetched."),
+ *     ),
+ *     "limit" = @ContextDefinition("integer",
+ *       label = @Translation("Limit"),
+ *       description = @Translation("Limit the maximum number of fetched entities."),
+ *       required = FALSE,
+ *     ),
+ *   },
+ *   provides = {
+ *      "entity_fetched" = @ContextDefinition("entity",
+ *        label = @Translation("Fetched entity"),
+ *        multiple = TRUE,
+ *      )
+ *   }
+ * )
+ *
+ * @todo: Add access callback information from Drupal 7.
+ */
+class EntityFetchByField extends RulesActionBase implements ContainerFactoryPluginInterface {
+
+  /**
+   * The entity manager.
+   *
+   * @var \Drupal\Core\Entity\EntityManagerInterface
+   */
+  protected $entityManager;
+
+  /**
+   * Constructs a EntityFetchByField object.
+   *
+   * @param array $configuration
+   *   A configuration array containing information about the plugin instance.
+   * @param string $plugin_id
+   *   The plugin ID for the plugin instance.
+   * @param mixed $plugin_definition
+   *   The plugin implementation definition.
+   * @param \Drupal\Core\Entity\EntityManagerInterface $entity_manager
+   *   The entity manager service.
+   */
+  public function __construct(array $configuration, $plugin_id, $plugin_definition, EntityManagerInterface $entity_manager) {
+    parent::__construct($configuration, $plugin_id, $plugin_definition);
+    $this->entityManager = $entity_manager;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container, array $configuration, $plugin_id, $plugin_definition) {
+    return new static(
+      $configuration,
+      $plugin_id,
+      $plugin_definition,
+      $container->get('entity.manager')
```

Comment:
we should inject the entity storage here by getting it from the entity manager, because we only use the storage.